ludwig/trainers/trainer.py

```diff
@@ -353,115 +366,62 @@ def train_for_tuning(
-                    logger.info(f"Explored learning_rate={current_learning_rate} loss={smoothed_loss}")
-
-                    # check whether loss is diverging
-                    if step_count > 0 and smoothed_loss > early_stop_threshold * best_loss:
-                        diverging = True
-                        break
-                    else:
-                        if smoothed_loss < best_loss or step_count == 0:
-                            best_loss = smoothed_loss
-
-                    # compute new learning rate
-                    if mode == "exponential":
-                        current_learning_rate = exponential_scheduler(current_learning_rate, step_count)
-                    else:
-                        current_learning_rate = linear_scheduler(current_learning_rate, step_count)
-
-                    self.set_optimizer_learning_rate(current_learning_rate)
-                    step_count += 1
+    def tune_learning_rate(self, config: ModelConfigDict) -> float:
+        """Recommend a learning rate based on the feautres in the configs and the associated encoders.
 
-                epoch += 1
+        param:
+            config: Ludwig config used to train the model.
+        """
+        logger.info("Tuning learning rate...")
 
-        optimal_lr = get_optimal_lr(losses, learning_rates)
-        if optimal_lr:
-            learning_rate = optimal_lr
+        features_with_large_encoders = {TEXT, IMAGE, SEQUENCE, AUTO, TIMESERIES}
+
+        encoders_in_config = set()
+        features_in_config = set()
+        trainable_in_config = set()
+        huggingface_encoders = set()
+
+        for item in get_huggingface_encoder_registry().values():
+            huggingface_encoders |= item.keys()
+
+        # input and output feature sections
+        for feature in config["input_features"] + config["output_features"]:
+            features_in_config.add(feature.get("type"))
+            feature_encoder = feature.get("encoder", {})
+            encoder_type = feature_encoder.get("type")
+            encoder_is_trainable = feature_encoder.get("trainable")
+            trainable_in_config.add(encoder_is_trainable)
+            encoders_in_config.add(encoder_type)
+
+        # default section
+        defaults = config.get("defaults", {})
+        for feature_type in defaults.keys() & features_in_config:
+            feature_defaults = defaults.get(feature_type, {})
+            encoder_section = feature_defaults.get("encoder", {})
+            encoder_type = encoder_section.get("type")
+            encoder_is_trainable = encoder_section.get("trainable")
+            trainable_in_config.add(encoder_is_trainable)
+            encoders_in_config.add(encoder_type)
+
+        is_feature_with_large_encoder = len(features_in_config & features_with_large_encoders) > 0
+        has_huggingface_encoder = len(encoders_in_config & huggingface_encoders) > 0
+        has_trainable_encoder = any(trainable_in_config)
```

Comment:
I would suggest changing this code a bit:

First, move these to separate helper functions in config_utils.py

```python
def has_trainable_encoder(config_dict: ModelConfigDict) -> bool:
    for feature in config_dict["input_features"]:
        feature_encoder = feature.get("encoder", {})
        encoder_type = feature_encoder.get("type")
        if encoder_section.get("trainable"):
            return True
    return False

def has_unstructured_input_feature(config_dict: ModelConfigDict) -> bool:
    for feature in config_dict["input_features"]:
        if feature.get("type", {}) in {TEXT, IMAGE, SEQUENCE, AUTO, TIMESERIES}:
            return True
    return False

def has_pretrained_encoder(config_dict: ModelConfigDict) -> bool:
    for feature in config_dict["input_features"]:
        feature_encoder = feature.get("encoder", {})
        encoder_type = feature_encoder.get("type")
        if encoder_type in get_huggingface_encoder_registry():
             return True
    return False
```

Then, in `learning_rate_utils.py` (or `learning_rate.py`), use these helpers to decide the learning rate, which I refactored a bit to [minimize nesting](https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html):

```python
def get_auto_learning_rate(config_dict: ModelConfigDict) -> float:
    """Uses config heuristics to determine an appropriate learning rate.

    The main idea behind the following heuristics is that smaller learning rates are more suitable for features with larger encoders, which are typically used with unstructured features.

    Note that these are meant to be rough heuristics that are solely based on feature types and the type of the corresponding encoder.

    More factors could be taken into consideration such as model size, dataset size, batch size, number of features, etc.
    """
    if not has_unstructured_input_feature(config_dict):
        return return 0.001

    if not has_pretrained_encoder(config_dict):
        return 0.0001

    if has_trainable_encoder(config_dict):
        return 0.00001
    return 0.00002
```